Corrected VM and container image size logic

zed-rishabh opened the pull request:
Update:
The issue is the image size and max size is incorrect in the volumemgr and zedmanager config/status for VMs. But while reporting the app metric back to the controller, sizes are correct because we are running and taking the output of the ```qemu-img```.
And for container images, EVE is not getting the image size in config from the controller so it is not handled and managed in EVE at all.
So with these changes, we are properly handling the container image size and correcting the VMs image max size in config and status.

zed-rishabh commented:
> 1. Is there an issue or other write-up (here in the PR comments itself would be fine too), what the problem is, how it manifested itself, and how this solves it?

The issue is image size and max size is incorrect in volumemgr and zedmanager config/status for VMs. But while reporting the app metric back to the controller, sizes are correct because we are running and taking the output of the ```qemu-img```.
And for container images, EVE is not getting the image size in config from the controller so it is not handled and managed in EVE at all.
So with these changes, we are properly handling the container image size and correcting the VMs image max size in config and status.

> 2. Does this conflict at all with the work @adarsh-zededa has been doing to separate out directory control between volumemgr and verifier?

I don't think it is related with the @adarsh-zededa work but he can confirm for the same.

deitch commented:
Thanks @zed-rishabh that helps.

I was confused as to why it would be zero, since we calculate it. But then I realized that we calculate it in _eve_, not in the _controller_, and we do it as part of the downloader update [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/zedUpload/ociutil/oci.go#L104) and [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/zedUpload/ociutil/oci.go#L155-L165), all of which is eve-side. Each [UpdateStats](https://github.com/lf-edge/eve/blob/master/pkg/pillar/zedUpload/ociutil/oci.go#L27-L32) contains the total expected filesize, but as you pointed out, that only happens when the download starts, and is eve-side.

Question: since we have the resolver that converts tag-to-sha when we first request to create a volume, even before the download starts, is there any reason _not_ to get size at that point too? For a container, it is pulling the manifest (most of the code for which already is there). For a VM image (or one that isn't in an OCI registry and therefore would have a manifest), we probably can get it too, depending on how the datastore implements it.

eriknordmark commented on pkg/pillar/cmd/volumemgr/handleconfig.go:
```diff
@@ -79,6 +80,13 @@ func vcCreate(ctx *volumemgrContext, objType string, key string,
+					errStr := fmt.Sprintf("Calculating size of container image failed: %v", err)
+					log.Error(errStr)
+				} else {
+					dos.MaxSizeBytes = uint64(info.Size())
```
You shouldn't need the MaxSizeBytes to do the snapshot.
We should track the actual size separately from the MaxSize; don't we have a field for this in VolumeStatus?

zed-rishabh replied:
Currently, we have two fields ```TargetSizeBytes``` and ```MaxSizeBytes``` in ```DownloadConfigOrigin``` in the VolumeStatus.
For container applications, both sizes are the same. And while creating container snapshots, we are not using sizes anywhere. It is only used for reporting purposes.

eriknordmark replied:
But if we want dos.MaxSizeBytes to be the size what we actually downloaded (as opposed to what we got from the MaxSizeBytes in the config), then we should apply that across the board and not just for containers.

zed-rishabh replied:
With these changes, we are applying it for both VM and containers.

eriknordmark replied:
OK

eriknordmark commented on pkg/pillar/cmd/volumemgr/initialvolumestatus.go:
```diff
@@ -144,20 +138,18 @@ func populateInitialVolumeStatus(ctx *volumemgrContext, dirName string) {
-			FileLocation: filelocation,
-			State:        types.CREATED_VOLUME,
-			// XXX Is this the correct size? vs. qcow2 size?
-			TargetSizeBytes: uint64(size),
```
TargetSizeBytes should track the current size (which we should extract from qemu-img for qcow2). Why leave it at zero here?

zed-rishabh replied:
While preparing initial status, we are not handling/filling sizes. At the time of volume creation, we are taking the sizes from the config.
Note: we are not taking the output of the qemu for filling TargetSizeBytes in master.

eriknordmark replied:
But volumemgr needs to be able to track the disk usage, and if we keep it at zero in the initial VolumeStatus then how will volumemgr track the disk usage?

zed-rishabh replied:
Initial status gets overwritten with the actual config passed from the zedmanager in volumemgr at the time volume creation.

eriknordmark replied:
OK

eriknordmark commented on pkg/pillar/cmd/volumemgr/updatestatus.go:
```diff
@@ -83,6 +85,20 @@ func doUpdate(ctx *volumemgrContext, status *types.VolumeStatus) (bool, bool) {
+			errStr := fmt.Sprintf("Calculating size of container image failed: %v", err)
+			log.Error(errStr)
+		} else {
+			status.DownloadOrigin.MaxSizeBytes = uint64(info.Size())
```
This is not the current size

zed-rishabh replied:
We are setting up the download size for containers here because it is 0 in config.

eriknordmark replied:
I don't want container-specific code. This adds more container specific code.

zed-rishabh replied:
We are not doing it separately. At the time of volume creation, if the initial status is available then we are preparing the snapshot again in volumemgr. So, at the same time we are also setting up the downloaded size also because the size is zero in the volume config from the zedmanager for the container applications.

eriknordmark replied:
My point is that we should not add any code in a container/isContainer specific path.
We need to make the code more uniform and this makes it less uniform across VMs and containers.

eriknordmark replied:
Thanks

eriknordmark commented on pkg/pillar/cmd/volumemgr/updatestatus.go:
```diff
@@ -163,6 +179,9 @@ func doDownload(ctx *volumemgrContext, status *types.VolumeStatus) bool {
 		changed = true
 	}
+	if status.DownloadOrigin.IsContainer {
+		status.DownloadOrigin.MaxSizeBytes = ds.Size
```
This is not the current size; status.TargetSizeBytes tracks the current size AFAIK

zed-rishabh replied:
We are setting up the download size for containers here because it is 0 in config. We can set ```TargetSizeBytes ``` equal to the download size for the containers.

eriknordmark replied:
OK

zed-rishabh commented:
>Question: since we have the resolver that converts tag-to-sha when we first request to create a volume, even before the download starts, is there any reason not to get size at that point too?

That's a very good point but the problem in filling the size while resolving the tag is that the sum of sizes in the manifest won't be equal to the size of the downloaded tar because we add some files in tar which are not in manifests at the time of download (eg. repositories, etc.). And, the size needs to be updated after the download again. So, I thought to update the size in the downloader here https://github.com/lf-edge/eve/blob/d624608545349aea55197505a928113bbb259203/pkg/pillar/cmd/downloader/syncop.go#L219

zed-rishabh commented:
> No description hence I have no idea what issue you found.

My bad. I updated the description. Thanks for pointing it out.

> When we find existing volumes it makes no sense to update dos.MaxSizeBytes, but status.TargetSizeBytes should be set (from size of directory, or from maxsize of a qcow2 file).

We are updating the ```dos.MaxSizeBytes``` only for containers at boot. And for containers, both MaxSizeBytes and TargetSizeBytes are the same. We can set TargetSizeBytes also.
Note: For VMs, MaxSizeBytes and TargetSizeBytes get filled from the config.

> Note that we are currently not reporting the size of the containers to the controllers, and we might not take them into account when figuring out the total amount of space used by the applications.

Yes, we added the functionality of reporting app metrics of container applications here https://github.com/lf-edge/eve/pull/1018

zed-rishabh commented:
>The Size we get from the controller is the Max Size (for downloads and the size of the volume). As such we use it e.g., to resize the qemu-img.

We get two sizes from the controller. One is the maximum download size and the second is the maximum volume size.

>But other than that parenthesis it doesn't make sense updating the max size internally.
We should update and report the current size (which is what you have in the other PR).

Maxsize is the maximum download size which we can set internally for containers after downloading it. Isn't it?
Note: With my changes, there is no change in VM size workflow except fixing ```TargetSizeBytes``` in zedmanager. These changes mostly for handling container image size.

deitch commented:
>> resolver

> sum of sizes in the manifest won't be equal to the size of the downloaded tar because we add some files in tar which are not in manifests at the time of download (eg. repositories, etc.). And, the size needs to be updated after the download again

True, although we can calculate it earlier as well. I also got the underlying library to report expected file size, so we can leverage that.

In any case, that shouldn't hold us up here. We can clean this up (as you are doing), and then later consider calculations at an earlier stage.

We also will be getting rid of the `repositories` file, and tar in general, once the [verifier PR](#1011) is in, as the next step, and will be downloading individual parts. That might mean we need to calculate size earlier anyways. 

Either way, it shouldn't hold this up.

eriknordmark commented:
> We are updating the `dos.MaxSizeBytes` only for containers at boot

That still seems wrong. Why shouldn't this come from zedcloud when the volume is created (where zero means unlimited)?

zed-rishabh commented:
@eriknordmark One of the major issues is fixed with these changes for VMs. Currently, volume resize is not happening to the specified size provided in the application manifest. The size of the volume is always equal to the downloaded size in the current master.

zed-rishabh commented:
> > We get two sizes from the controller. One is the maximum download size and the second is the maximum volume size.
> > >
> 
> You mean with the new Volume API? The old Drive only has one size AFAIK

No, we get two sizes in the old drive as well.
